docs/design.md

```diff
@@ -1,87 +1,133 @@
-
-Therefore Kubernetes AWS volume plugin maintains cache of used / free device names for each node. This cache is lost when controller-manager process restarts. We try to populate the cache during startup, however there are some corner cases when this fails. TODO: exact flow how we can get wrong cache.
-
-It would be great if either AWS itself assigned the device names, or there would be robust way how to restore the cache after restart, e.g. using some persistent database. Kubernetes should not care about the device names at all.
-
-### DescribeVolumes quota
-
-In order to attach/detach volume to/from a node, current AWS cloud provider issues AWS.AttachVolume/DetachVolume call and then it polls DescribeVolume until the volume is attached or detached. The frequency of DescribeVolume is quite high to minimize delay between AWS finishing attachment of the volume and Kubernetes discovering that. Sometimes we even hit API quota for these calls.
-
-It would be better if CSI driver could get reliable and fast event from AWS when a volume has become attached / detached.
-
-Or the driver could batch the calls and issue one big DescribeVolume call with every volume that&#39;s being attached/detached in it.
-
-### AWS API weirdness
-
-AWS API is quite different to all other clouds.
-
-- AWS.AttachVolume/DetachVolume can take ages (days, weeks) to complete. For example, when Kubernetes tries to detach a volume that&#39;s still mounted on a node, it will be Detaching until the volume is unmounted or Kubernetes issues force-detach. All other clouds return sane error relatively quickly (e.g. &quot;volume is in use&quot;) or force-detach the volume.
-- AWS.DetachVolume with force-detach is useless. Documentation says: Forced detachment of a stuck volume can cause damage to the file system or the data it contains or an inability to attach a new volume using the same device name, unless you reboot the instance.
-  - We cannot reboot instance after each force-detach nor we can afford to &quot;loose&quot; a device name. AWS supports only 40 volumes per node and even that is quite low number already.
-- AWS.CreateVolume is not idempotent. There is no way to create a volume with ID provided by user. Such call would then fail when such volume already exists.
-- AWS.CreateVolume does not return errors when creating an encrypted volume using either non-existing or non-available KMS key (e.g. with wrong permission). It returns success instead and it even returns volumeID of some volume! This volume exists for a short while and it&#39;s deleted in couple of seconds.
 
-### Errors with slow kubelet
-
-Very rarely a node gets too busy and kubelet starves for CPU. It does not unmount a volume when it should and Kubernetes initiates detach of the volume.
+## EBS CSI Driver on Kubernetes
+
+### High Level Summary
```

Comment:
This section is dangerously close to documenting CSI and/or CSI sidecar design rather than EBS CSI design. For example, if the information here can be found in the CSI spec's architecture section (https://github.com/container-storage-interface/spec/blob/master/spec.md#architecture) we are better off linking there than making our own version.